### ISSUE TITLE
Training details of shepherd.

### ISSUE THREAD

**Issue by VicFisher** [coleygroup/shepherd#4]
Hello, it's really a wonderful project. 
I am recently trying to re-train the proposed model. Would you please explain the following training details:

**(1) Validation Sets**: I noticed that there doesn't appear to be a validation set during training for the Trainer. How is the model evaluated throughout the training process? Additionally, is the final inference checkpoint taken from the last model checkpoint?

**(2) Model Convergence**: What criteria do you use to determine whether the model has converged? Specifically, could you provide information on the expected range of losses for each component of the model?
Thank you for your assistance, and I look forward to your response.

Best regards.
Vic

**Comment by kentoabeywardane** [comment]
Hi Vic,
Thank you for taking interest in our model. Here are our explanations:

**(1) Validation sets**: We did not use validation sets during training since we were not worried about overfitting as the dataset size was so large and had a low number of epochs (~14 epochs). Furthermore, because the losses from diffusion are so noisy, they are difficult to interpret. Rather, manual checks were done sporadically for the validity rate of unconditionally generated molecules (i.e., ensure improvement in RDKit validity and visual inspection of generated molecules). 
Yes, the final inference checkpoint uses the last model checkpoint since we did not use early stopping.

Note that the training process may benefit from incorporating different training protocols --in addition to hyperparameter tuning-- such as LR decay, EMA, or adding in a validation set where conditional evaluations could be carried out by 3D similarity scoring functions. This was not the focus of our experiments due to limited computational budget.

**(2) Model convergence**: We did not use losses to check for convergence since they reach a noisy plateau which was found to be unreliable as there was improvement in performance after the plateau. Instead, we manually checked whether generated molecules attained good validity as mentioned in (1). 

Please let me know if you have some further questions!

Best,
Kento

**Comment by VicFisher** [comment]
Thank you for your quick response. Your explanation has totally solve my problem. 发自我的手机-------- 原始邮件 --------发件人： Kento Abeywardane ***@***.***>日期： 2025年1月5日周日 07:03收件人： coleygroup/shepherd ***@***.***>抄送： VicFisher ***@***.***>, Author ***@***.***>主    题： Re: [coleygroup/shepherd] Training details of shepherd. (Issue #4)
Hi Vic,
Thank you for taking interest in our model. Here are our explanations:
(1) Validation sets: We did not use validation sets during training since we were not worried about overfitting as the dataset size was so large and had a low number of epochs (~14 epochs). Furthermore, because the losses from diffusion are so noisy, they are difficult to interpret. Rather, manual checks were done sporadically for the validity rate of unconditionally generated molecules (i.e., ensure improvement in RDKit validity and visual inspection of generated molecules).
Yes, the final inference checkpoint uses the last model checkpoint since we did not use early stopping.
Note that the training process may benefit from incorporating different training protocols --in addition to hyperparameter tuning-- such as LR decay, EMA, or adding in a validation set where conditional evaluations could be carried out by 3D similarity scoring functions. This was not the focus of our experiments due to limited computational budget.
(2) Model convergence: We did not use losses to check for convergence since they reach a noisy plateau which was found to be unreliable as there was improvement in performance after the plateau. Instead, we manually checked whether generated molecules attained good validity as mentioned in (1).
Please let me know if you have some further questions!
Best,
Kento
